libia2/threads.c

```diff
@@ -17,12 +18,54 @@ void *ia2_thread_begin(void *arg) {
   /* Free the thunk. */
   munmap(arg, sizeof(struct ia2_thread_thunk));
 
-  init_stacks();
   protect_tls();
+  init_stacks();
 
-  /* TODO: switch to compartment stack before calling `fn` */
+  /* Determine the current compartment so know which stack to use. */
+  uint32_t pkru = 0;
+  __asm__ volatile(
+      /* clang-format off */
+      "xor %%ecx,%%ecx\n"
+      IA2_RDPKRU "\n"
+      /* clang-format on */
+      : "=a"(pkru)::"ecx", "edx");
+  /* returns NULL in insecure mode, because we can't read a pkru there. */
+  void **new_sp_addr = ia2_stackptr_for_pkru(pkru);
```

Comment:
assert on null?

Reply:
This is more of an explanation of why we do not use the result as a stack pointer in insecure mode than a premise that anything else relies on. We could add an assert, but there's no interesting logic there for it to validate.